src/librustc_mir/transform/check_consts/validation.rs

```diff
@@ -341,27 +348,92 @@ impl Visitor<'tcx> for Validator<'_, 'mir, 'tcx> {
                         MutatingUseContext::Borrow,
                     ),
                 };
-                self.visit_place_base(&place.base, ctx, location);
-                self.visit_projection(&place.base, proj, ctx, location);
-            } else {
-                self.super_rvalue(rvalue, location);
+                self.visit_place_base(&borrowed_place.base, ctx, location);
+                self.visit_projection(&borrowed_place.base, inner, ctx, location);
+                return;
             }
-        } else {
-            self.super_rvalue(rvalue, location);
-        }
 
-        match *rvalue {
-            Rvalue::Use(_) |
-            Rvalue::Repeat(..) |
-            Rvalue::UnaryOp(UnOp::Neg, _) |
-            Rvalue::UnaryOp(UnOp::Not, _) |
-            Rvalue::NullaryOp(NullOp::SizeOf, _) |
-            Rvalue::CheckedBinaryOp(..) |
-            Rvalue::Cast(CastKind::Pointer(_), ..) |
-            Rvalue::Discriminant(..) |
-            Rvalue::Len(_) |
-            Rvalue::Ref(..) |
-            Rvalue::Aggregate(..) => {}
+            // It is *always* legal to borrow a `static` from another `static`, even if it contains
+            // interior mutability. Accessing a `static` from a `const` is disallowed by other
+            // checks.
+            | Rvalue::Ref(_, BorrowKind::Shared, ref borrowed_place)
+            | Rvalue::Ref(_, BorrowKind::Shallow, ref borrowed_place)
+            | Rvalue::Ref(_, BorrowKind::Unique, ref borrowed_place)
+            if matches!(borrowed_place.base, PlaceBase::Static(_))
+            => {}
+
+            // Check for shared borrows of values with interior mutability.
+            | Rvalue::Ref(_, kind @ BorrowKind::Shared, ref borrowed_place)
+            | Rvalue::Ref(_, kind @ BorrowKind::Shallow, ref borrowed_place)
+            | Rvalue::Ref(_, kind @ BorrowKind::Unique, ref borrowed_place)
```

Comment:
Unique borrows should be qualified the same way as mutable borrows.

Reply:
This would be a change in behavior, since `HasMutInterior` returns `true` only for `BorrowKind::Mut` at the moment. Looking at the docs, it appears that these are only generated for closure captures of variables of type `&mut _`, so this change wouldn't break any code right?

Reply:
That sounds correct